api.c

```diff
@@ -0,0 +1,1055 @@
+                    break;
+            }
+        } else
+        if (c == ' ' || c == '\t') { // Whitespace
+            switch(curr_state) {
+                case InArg:
+                    push_back(&out_args, curr_arg);
+                    curr_state = OutOfArg;
+                    break;
+                case InArgQuote:
+                    str_push_c(&curr_arg, c);
+                    break;
+                case OutOfArg:
+                    break;
+            }
+        } else {
+            switch(curr_state) {
+                case InArg:
+                case InArgQuote:
+                    str_push_c(&curr_arg, c);
+                    break;
+                case OutOfArg:
+                    free(curr_arg.value);
+                    init_str(&curr_arg, 20);
+                    str_push_c(&curr_arg, c);
+                    curr_state = InArg;
+                    break;
+            }
+        };
+        i++;
+    }
+
+    if (curr_state == InArg) {
+        push_back(&out_args, curr_arg);
+    } else if (curr_state == InArgQuote) {
+        return 1;
+    }
+
+    size_t argc = out_args.size;
+    char **argv = str_destruct(out_args);
+
+    free(curr_arg.value);
+    free(out_args.items);
+
+    data->argc = argc;
+    data->argv = argv;
+
+    return 0;
+}
+
+void *listen_api(void* _) {
+
+    if (access(api_fifo_path, F_OK) == 0) {
+        if (remove(api_fifo_path) != 0) {
+            fprintf(stderr, "Could not remove file \"%s\": %s\n",
+                    api_fifo_path, strerror(errno));
+        }
+    }
+    if (mkfifo(api_fifo_path, S_IRWXU) != 0) {
+        fprintf(stderr, "Could not create FIFO file \"%s\": %s\n",
+                api_fifo_path, strerror(errno));
+    }
+
+    FILE *fifo;
+    char c;
+    int i;
+    int times = 0;
+    char *text = (char *) malloc(50);
+    arg_data data;
+
+    DEBUG("Created fifo at \"%s\"\n", api_fifo_path);
+
+    while (1) {
+        strcpy(text, "command ");
+        i = strlen(text);
+
+        fifo = fopen(api_fifo_path, "rb");
+        if (fifo == NULL) {
+            fprintf(stderr, "Could not open FIFO file \"%s\": %s\n",
+                    api_fifo_path, strerror(errno));
+            break;
+        }
+        while ((c = fgetc(fifo)) != EOF) {
+            if (sizeof(text) <= i) {
+                text = (char *) realloc(text, i * 2);
+            }
+            if (c == '\n') {
+                text[i] = ' ';
+            } else {
+                text[i] = c;
+            }
+            i++;
+        }
+        text[i] = '\0';
+        fclose(fifo);
+
+        DEBUG("API message received - %d\n", times);
+        DEBUG("Message: \"%s\"\n", text);
+        times++;
+        if (parse_args(&data, text) != 0) {
+            fprintf(stderr, "Couldn not parse message\n");
+            continue;
+        }
+
+        update_arguments(data.argc, data.argv);
+        init_colors_once();
+
+        free(text);
+        text = (char *) malloc(50);
```

Comment:
what's this malloc doing here?

Reply:
Reallocate the memory to nothing? As `free` deallocates the previous allocated memory for **text**, I though it might be needed to allocate **text** again without keeping any increased memory.

If this seems bad, any suggestions?

Reply:
You can leave text to go out of scope once you free it, or set it to NULL if you want. Malloc like this is just memory leaks.

Oh, and you don't need to cast the result of malloc since it returns void pointers (and shouldn't https://stackoverflow.com/a/605858/6023997).